### PR TITLE
*: Anonymous field for RegionInfo - part 1

### PR DESCRIPTION
**Pull request by nolouch**
## What have you changed? (required)

We only read the RegionInfo from the cache. For more clear and remove useless clone usage, make the RegionInfo read-only.
## What are the type of the changes (required)?

- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested (required)?

unit tests.
## Refer to a related PR or issue link (optional)
https://github.com/pingcap/pd/pull/1205

### CODE REVIEW

**Comment by disksing** [comment]
CI failed. @nolouch

**Comment by nolouch** [comment]
@disksing  the test fixed in part3.

**Comment by disksing** [comment]
em, you need to pick the fix to this branch. We can't merge it into master before tests passed.

**Comment by nolouch** [comment]
@disksing part 3 not send to the master branch, it submitted to branch `part 2`. and `part 2`  submitted to `part 1`, only part 1 for the master. so you can review them one by one.

File: server/core/region_option.go [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+}
+
+// SetWrittenBytes sets the written bytes for the region.
+func SetWrittenBytes(v uint64) RegionCreateOption {
+	return func(region *RegionInfo) {
+		region.writtenBytes = v
+	}
+}
+
+// WithRemoveStorePeer remove the specified peer for the region.
```
Comment (rleungx) [inline server/core/region_option.go]: removes

File: server/core/region_option.go [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+
+// WithPromoteLearner promotes the learner.
+func WithPromoteLearner(peerID uint64) RegionCreateOption {
+	return func(region *RegionInfo) {
+		for _, p := range region.GetPeers() {
+			if p.GetId() == peerID {
+				p.IsLearner = false
+			}
+		}
+		for _, l := range region.learners {
```
Comment (rleungx) [inline server/core/region_option.go]: Do we need this?

**Comment by disksing** [comment]
@nolouch After other parts are merged to part1, part1 will become a huge PR which is hard to review. I think it would be better to merge PRs to master one by one.

**Comment by disksing** [comment]
LGTM

File: server/core/region.go [hunk 3]
```diff
@@ -232,39 +229,17 @@ func (r *RegionInfo) GetStoreVoter(storeID uint64) *metapb.Peer {
...
-		r.Learners = append(r.Learners, peer)
-	} else {
-		r.Voters = append(r.Voters, peer)
-	}
-}
-
 // GetStoreIds returns a map indicate the region distributed.
 func (r *RegionInfo) GetStoreIds() map[uint64]struct{} {
-	peers := r.GetPeers()
+	peers := r.meta.GetPeers()
```
Comment (rleungx) [inline server/core/region.go]: It seems we can use `GetPeers` directly?

File: server/core/region.go [hunk 4]
```diff
@@ -330,6 +370,18 @@ type RegionStat struct {
...
 }
 
+// NewRegionStat returns a RegionStat.
+func NewRegionStat(region *RegionInfo, flowBytes uint64, antiCount int) *RegionStat {
+	return &RegionStat{
+		RegionID:       region.GetID(),
+		FlowBytes:      flowBytes,
+		LastUpdateTime: time.Now(),
+		StoreID:        region.leader.GetStoreId(),
+		Version:        region.meta.GetRegionEpoch().GetVersion(),
```
Comment (rleungx) [inline server/core/region.go]: same as `GetPeers`

File: server/core/region.go (hunk 3 above)
Comment (nolouch) [reply server/core/region.go]: I want to change the export function `GetPeers` without getting the pointer of the origin peers in the next PR. now we got the peer can also change the read-only  `RegionInfo`. So, in the inner, I still use the method of meta.

**Comment by nolouch** [comment]
PTAL @disksing @rleungx

File: server/api/trend_test.go [hunk 5]
```diff
@@ -38,39 +38,38 @@ func (s *testTrendSuite) TestTrend(c *C) {
...
 	newPeerID := op.Step(0).(schedule.AddLearner).PeerID
-	region := s.newRegionInfo(5, "a", "b", 3, 2, []uint64{1, 2}, []uint64{3}, 1)
-	region.Learners[0].Id = newPeerID
-	mustRegionHeartbeat(c, svr, region)
-	region = s.newRegionInfo(5, "a", "b", 4, 2, []uint64{1, 3}, nil, 1)
-	region.Peers[1].Id = newPeerID
-	mustRegionHeartbeat(c, svr, region)
+	region5 = region5.Clone(core.WithAddPeer(&metapb.Peer{Id: newPeerID, StoreId: 3, IsLearner: true}))
+	mustRegionHeartbeat(c, svr, region5)
+	region5 = region5.Clone(core.WithPromoteLearner(newPeerID), core.WithRemoveStorePeer(2))
```
Comment (rleungx) [inline server/api/trend_test.go]: Does `confVer` need to be incremented?

**Pull request merged**